File: clients/csharp/DebugInterface.cs
using System.IO;

namespace AiCup22
{
    public class DebugInterface
    {
        private BinaryWriter writer;
        private BinaryReader reader;
        public DebugInterface(BinaryReader reader, BinaryWriter writer)
        {
            this.reader = reader;
            this.writer = writer;
        }

        public void AddPlacedText(AiCup22.Model.Vec2 position, string text, AiCup22.Model.Vec2 alignment, double size, AiCup22.Debugging.Color color)
        {
            Add(new AiCup22.Debugging.DebugData.PlacedText(position, text, alignment, size, color));
        }
        
        public void AddCircle(AiCup22.Model.Vec2 position, double radius, AiCup22.Debugging.Color color)
        {
            Add(new AiCup22.Debugging.DebugData.Circle(position, radius, color));
        }
        
        public void AddGradientCircle(AiCup22.Model.Vec2 position, double radius, AiCup22.Debugging.Color innerColor, AiCup22.Debugging.Color outerColor)
        {
            Add(new AiCup22.Debugging.DebugData.GradientCircle(position, radius, innerColor, outerColor));
        }
        
        public void AddRing(AiCup22.Model.Vec2 position, double radius, double width, AiCup22.Debugging.Color color)
        {
            Add(new AiCup22.Debugging.DebugData.Ring(position, radius, width, color));
        }
        
        public void AddPie(AiCup22.Model.Vec2 position, double radius, double startAngle, double endAngle, AiCup22.Debugging.Color color)
        {
            Add(new AiCup22.Debugging.DebugData.Pie(position, radius, startAngle, endAngle, color));
        }
        
        public void AddArc(AiCup22.Model.Vec2 position, double radius, double width, double startAngle, double endAngle, AiCup22.Debugging.Color color)
        {
            Add(new AiCup22.Debugging.DebugData.Arc(position, radius, width, startAngle, endAngle, color));
        }
        
        public void AddRect(AiCup22.Model.Vec2 bottomLeft, AiCup22.Model.Vec2 size, AiCup22.Debugging.Color color)
        {
            Add(new AiCup22.Debugging.DebugData.Rect(bottomLeft, size, color));
        }
        
        public void AddPolygon(AiCup22.Model.Vec2[] vertices, AiCup22.Debugging.Color color)
        {
            Add(new AiCup22.Debugging.DebugData.Polygon(vertices, color));
        }
        
        public void AddGradientPolygon(AiCup22.Debugging.ColoredVertex[] vertices)
        {
            Add(new AiCup22.Debugging.DebugData.GradientPolygon(vertices));
        }
        
        public void AddSegment(AiCup22.Model.Vec2 firstEnd, AiCup22.Model.Vec2 secondEnd, double width, AiCup22.Debugging.Color color)
        {
            Add(new AiCup22.Debugging.DebugData.Segment(firstEnd, secondEnd, width, color));
        }
        
        public void AddGradientSegment(AiCup22.Model.Vec2 firstEnd, AiCup22.Debugging.Color firstColor, AiCup22.Model.Vec2 secondEnd, AiCup22.Debugging.Color secondColor, double width)
        {
            Add(new AiCup22.Debugging.DebugData.GradientSegment(firstEnd, firstColor, secondEnd, secondColor, width));
        }
        
        public void AddPolyLine(AiCup22.Model.Vec2[] vertices, double width, AiCup22.Debugging.Color color)
        {
            Add(new AiCup22.Debugging.DebugData.PolyLine(vertices, width, color));
        }
        
        public void AddGradientPolyLine(AiCup22.Debugging.ColoredVertex[] vertices, double width)
        {
            Add(new AiCup22.Debugging.DebugData.GradientPolyLine(vertices, width));
        }

        public void Add(AiCup22.Debugging.DebugData debugData)
        {
            Send(new AiCup22.Debugging.DebugCommand.Add(debugData));
        }
        
        public void Clear()
        {
            Send(new AiCup22.Debugging.DebugCommand.Clear());
        }
        
        public void SetAutoFlush(bool enable)
        {
            Send(new AiCup22.Debugging.DebugCommand.SetAutoFlush(enable));
        }
        
        public void Flush()
        {
            Send(new AiCup22.Debugging.DebugCommand.Flush());
        }

        public void Send(AiCup22.Debugging.DebugCommand command)
        {
            new AiCup22.Codegame.ClientMessage.DebugMessage(command).WriteTo(writer);
            writer.Flush();
        }

        public AiCup22.Debugging.DebugState GetState()
        {
            new AiCup22.Codegame.ClientMessage.RequestDebugState().WriteTo(writer);
            writer.Flush();
            return AiCup22.Debugging.DebugState.ReadFrom(reader);
        }
    }
}

File: clients/rust/src/debug_interface.rs
use ai_cup_22::*;

pub struct DebugInterface<'a> {
    reader: &'a mut dyn std::io::Read,
    writer: &'a mut dyn std::io::Write,
}

impl<'a> DebugInterface<'a> {
    pub fn new(reader: &'a mut dyn std::io::Read, writer: &'a mut dyn std::io::Write) -> Self {
        Self { reader, writer }
    }

    pub fn add_placed_text(&mut self, position: model::Vec2, text: String, alignment: model::Vec2, size: f64, color: debugging::Color) {
        self.add(debugging::DebugData::PlacedText { position, text, alignment, size, color });
    }
    
    pub fn add_circle(&mut self, position: model::Vec2, radius: f64, color: debugging::Color) {
        self.add(debugging::DebugData::Circle { position, radius, color });
    }
    
    pub fn add_gradient_circle(&mut self, position: model::Vec2, radius: f64, inner_color: debugging::Color, outer_color: debugging::Color) {
        self.add(debugging::DebugData::GradientCircle { position, radius, inner_color, outer_color });
    }
    
    pub fn add_ring(&mut self, position: model::Vec2, radius: f64, width: f64, color: debugging::Color) {
        self.add(debugging::DebugData::Ring { position, radius, width, color });
    }
    
    pub fn add_pie(&mut self, position: model::Vec2, radius: f64, start_angle: f64, end_angle: f64, color: debugging::Color) {
        self.add(debugging::DebugData::Pie { position, radius, start_angle, end_angle, color });
    }
    
    pub fn add_arc(&mut self, position: model::Vec2, radius: f64, width: f64, start_angle: f64, end_angle: f64, color: debugging::Color) {
        self.add(debugging::DebugData::Arc { position, radius, width, start_angle, end_angle, color });
    }
    
    pub fn add_rect(&mut self, bottom_left: model::Vec2, size: model::Vec2, color: debugging::Color) {
        self.add(debugging::DebugData::Rect { bottom_left, size, color });
    }
    
    pub fn add_polygon(&mut self, vertices: Vec<model::Vec2>, color: debugging::Color) {
        self.add(debugging::DebugData::Polygon { vertices, color });
    }
    
    pub fn add_gradient_polygon(&mut self, vertices: Vec<debugging::ColoredVertex>) {
        self.add(debugging::DebugData::GradientPolygon { vertices });
    }
    
    pub fn add_segment(&mut self, first_end: model::Vec2, second_end: model::Vec2, width: f64, color: debugging::Color) {
        self.add(debugging::DebugData::Segment { first_end, second_end, width, color });
    }
    
    pub fn add_gradient_segment(&mut self, first_end: model::Vec2, first_color: debugging::Color, second_end: model::Vec2, second_color: debugging::Color, width: f64) {
        self.add(debugging::DebugData::GradientSegment { first_end, first_color, second_end, second_color, width });
    }
    
    pub fn add_poly_line(&mut self, vertices: Vec<model::Vec2>, width: f64, color: debugging::Color) {
        self.add(debugging::DebugData::PolyLine { vertices, width, color });
    }
    
    pub fn add_gradient_poly_line(&mut self, vertices: Vec<debugging::ColoredVertex>, width: f64) {
        self.add(debugging::DebugData::GradientPolyLine { vertices, width });
    }

    pub fn add(&mut self, debug_data: debugging::DebugData) {
        self.send(debugging::DebugCommand::Add { debug_data });
    }
    
    pub fn clear(&mut self) {
        self.send(debugging::DebugCommand::Clear {  });
    }
    
    pub fn set_auto_flush(&mut self, enable: bool) {
        self.send(debugging::DebugCommand::SetAutoFlush { enable });
    }
    
    pub fn flush(&mut self) {
        self.send(debugging::DebugCommand::Flush {  });
    }

    pub fn send(&mut self, command: debugging::DebugCommand) {
        use trans::Trans;
        codegame::ClientMessage::DebugMessage { command }
            .write_to(self.writer)
            .expect("Failed to write custom debug data");
        self.writer.flush().expect("Failed to flush");
    }

    pub fn get_state(&mut self) -> debugging::DebugState {
        use trans::Trans;
        codegame::ClientMessage::RequestDebugState {}
            .write_to(self.writer)
            .expect("Failed to write request debug state message");
        self.writer.flush().expect("Failed to flush");
        debugging::DebugState::read_from(self.reader).expect("Failed to read debug state")
    }
}